src/devices/watts_thermostat.c

```diff
@@ -0,0 +1,199 @@
+     [ 1] count:   28,  width:  576 us [576;576]    (1151 S)
+    Pulse period distribution:
+     [ 0] count:    1,  width: 291328 us [291328;291328]    (582657 S)
+     [ 1] count:   15,  width: 1206 us [1205;1216]  (2412 S)
+     [ 2] count:   14,  width:  522 us [520;530]    (1043 S)
+     [ 3] count:   24,  width:  864 us [862;877]    (1727 S)
+    Pulse timing distribution:
+     [ 0] count:    1,  width: 291096 us [291096;291096]    (582191 S)
+     [ 1] count:   54,  width:  602 us [576;640]    (1204 S)
+     [ 2] count:   54,  width:  262 us [232;302]    ( 523 S)
+     [ 3] count:    1,  width: 100000 us [100000;100000]    (200001 S)
+    Level estimates [high, low]:  15961,    150
+    RSSI: -0.1 dB SNR: 20.3 dB Noise: -20.4 dB
+    Frequency offsets [F1, F2]:   17869,      0     (+545.3 kHz, +0.0 kHz)
+    Guessing modulation: Pulse Width Modulation with sync/delimiter
+    view at https://triq.org/pdv/#AAB104FFFF025A0105FFFF8291A291A1A291A29291A29291A291A1A1A2929291A29291A1A1A1A1A29291A291A1A1A1A291A1A29291A1A29291A29291A1A2929291A355
+    Attempting demodulation... short_width: 288, long_width: 630, reset_limit: 577, sync_width: 291096
+    Use a flex decoder with -X 'n=name,m=OOK_PWM,s=288,l=630,r=577,g=0,t=0,y=291096'
+    pulse_slicer_pwm(): Analyzer Device
+    bitbuffer:: Number of rows: 1
+    [00] {54} 5a 4b 89 f2 f6 64 c4
+
+
+Data Layout:
+
+    10100101   1011010001110110   1000   100100001   000011000   10101011
+    preamble   id                 flags  temp         setpoint   chksum
+
+    All fields need reflection, possibly easier to just reverse the whole
+    row first. The only flag found is PAIRING (0b0001). Chksum seems to be
+    additive.
+
+Raw data:
+
+    {54}5ab24971f79994
+    {54}5ab24971f79994
+    {54}5ab249f1f79b94
+    {54}5ab249f1f79b94
+    {54}5ab249f9f79854
+    {54}5ab249f5f79a54
+    {54}5ab249f68f998c
+    {54}5ab249f98f9a4c
+    {54}5ab249f58b9a4c
+    {54}5ab249fb8f9acc
+
+    https://tinyurl.com/2z5jtfuu
+
+    Format string:
+    ID:^16d FLAGS:^4b TEMP:^9d SETP:^9d CHK:^8d
+
+Decoded example:
+
+    ID:28205 FLAGS:0001 TEMP:265 SETP:048 CHK:214
+
+*/
+
+static int watts_thermostat_decode(r_device *decoder, bitbuffer_t *bitbuffer)
+{
+    int ret                          = 0;
+
+    if (bitbuffer->num_rows != 1) {
+        decoder_logf(decoder, 2, __func__, "Only expect a single row. num_rows=%d", bitbuffer->num_rows);
+        ret = DECODE_ABORT_EARLY;
+        return ret;
+    }
+
+    uint8_t const preamble_pattern = 0xa5;
+
+    for (unsigned row = 0; row < bitbuffer->num_rows; ++row) {
+        // we expect 54 bits
+        if (bitbuffer->bits_per_row[row] != 54) {
+            decoder_log(decoder, 2, __func__, "Length check fail");
+            ret = DECODE_ABORT_LENGTH;
+            continue;
+        }
+
+        bitbuffer_invert(bitbuffer);
+
+        unsigned pos = bitbuffer_search(bitbuffer, row, 0, &preamble_pattern, 8);
```

Comment:
If we are looking for a preamble, why are we strict on length?   Preamble tends to mean that once found we decode so some stray bits before are not such an issue.   I'm not saying this is wrong, but it would be good to have rationale.,

Reply:
Not really sure since this I'm pretty new to RF decoding. But is the preamble enough to be sure we detect the correct protocol? It's not like this is regulated in any way and any manufacturer is free to make a protocol that starts with those same bits. Checking the length in addition to the preamble is at least more unique. Also, since we check the length, we won't need to do any range checking at the rest of the `bitbuffer_search()` calls (have not checked whether that is necessary).

Reply:
I may be confusing preamble with synchronization.   There are various ways for a receiver to identify the first bit.   The question is how that first bit of preamble is lined up; there may be sync first, and then it's a "identify the device type" word.

Certainly you should validate everything you can including length.  I was questioning "always N and preamble starts" vs "after you find the preamble, N bits including preamble to end".   So basically, either you ask "are the first bits the preamble" and expect fixed buffer size, or you search, and it's ok to have junk bits before but you need the right number afterwards.  I can't tell from the code what the plan is and why.   I, more than others, want the reader to be led to understand, partly because I try to leave such notes for future self.  Sometimes I have read code I wrote 20 years ago.

Reply:
I'll start with a search for the preamble and do a minimum length check from that position (any additional tail bits would then be ignored). However I'm thinking I can hardcode that it expects only one row, is that okay?